src/win/util.c

```diff
@@ -484,7 +484,7 @@ uint64_t uv__hrtime(double scale) {
    * performance counter interval, integer math could cause this computation
    * to overflow. Therefore we resort to floating point math.
    */
-  return (uint64_t) ((double) counter.QuadPart * hrtime_interval_ * scale);
+  return (uint64_t) (counter.QuadPart * (uint64_t) scale / hrtime_interval_);
```

Comment:
QPC wrapping is not the issue I pointed out. The issue is QPC overflowing a double's precision, and the imprecise math that libuv does with it on top of that:

- the result of QPC can have more significant bits than a double can hold, which means the first cast in the original code can cause loss of precision. An interval of 1s already uses 32 bits, in a little less than 2 weeks the int64 can no longer fit in a double without losing least significant digits.

- hrtime_interval_ is obtained by inverting an int64 and storing it in a double, again losing precision.

That's why I suggested to move the double cast to the end of the computation - `(counter.QuadPart / hrtime_interval_) * scale`

----

The proper solution for measuring time deltas with an HR timer would be to ignore process.hrtime() entirely, and instead have two sets of API like `var hrtimer = new HRTimer(); hrtimer.start()` and `hrtimer.stop()`. The HRTimer instance could internally maintain the int64 ticks value. stop() could then do the subtraction and do the same FP math to return `[seconds, nanoseconds]`. This way you only require the delta to fit in a double's bounds (2^52 ns = 2 months - no one should run a timer that long).

process.hrtime() is flawed in that it does not return the original ticks value. If you assume that the zero of QPC is when the machine booted, and that an interval of 1s requires 32 digits of precision, then after 2 months of uptime the number of nanoseconds can no longer fit in a double without losing precision, and you can have two calls to process.hrtime() a few nanoseconds apart return you the same value. (Of course it *can't* return the ticks value because JS has no 64-bit integers anyway.)